### PR TITLE
Docling models path

### PR DESCRIPTION
**Pull request by aakankshaduggal**
#335

### CODE REVIEW

File: src/instructlab/sdg/generate_data.py [hunk 1]
```diff
@@ -210,15 +210,28 @@ def _context_init(
...
     # Search for the pipeline in User and Site data directories
     # then for a package defined pipeline
     # and finally pipelines referenced by absolute path
-    data_dirs = [os.path.join(xdg_data_home(), "instructlab", "sdg")]
-    data_dirs.extend(os.path.join(dir, "instructlab", "sdg") for dir in xdg_data_dirs())
+    data_dir = os.path.join(xdg_data_home(), "instructlab", "sdg", "models", "docling")
+    data_dirs = [data_dir]
+    data_dirs.extend(
+        os.path.join(dir, "instructlab", "sdg", "models", "docling") for dir in xdg_data_dirs()
+    )
```
Comment (jaideepr97) [inline src/instructlab/sdg/generate_data.py:223]: changing the base data dir to be `instructlab/sdg/models/docling` is probably not a good idea. you still want to look for all other assets under `instructlab/sdg` the way it currently works

I would just try and replicate the existing code flow for how we read other assets now and extend it to look for `models/docling-models`

File: src/instructlab/sdg/generate_data.py [hunk 2]
```diff
@@ -210,15 +210,28 @@ def _context_init(
...
+
+    # Set `docling_model_path` to consistently use `data_dir`
+    docling_model_path = Path(data_dir)
+    os.makedirs(docling_model_path, exist_ok=True)
+
+    if not os.listdir(docling_model_path):
+        # Download models if directory is empty
+        logger.info("Docling models for chunking not found locally. Downloading from Hugging Face...")
+        StandardPdfPipeline.download_models_hf()
```
Comment (jaideepr97) [inline src/instructlab/sdg/generate_data.py:232]: I would suggest moving this as close to where docling code gets executed as possible - rather than right out of the gate here

File: src/instructlab/sdg/utils/chunkers.py [hunk 3]
```diff
@@ -211,8 +215,11 @@ def chunk_documents(self) -> List:
         if self.document_paths == []:
             return []
 
-        model_artifacts_path = StandardPdfPipeline.download_models_hf()
-        pipeline_options = PdfPipelineOptions(artifacts_path=model_artifacts_path)
+        if not self.docling_model_path.exists():
+            raise FileNotFoundError(f"Docling model path not found: {self.docling_model_path}")
+        print("docling_model_path", docling_model_path)
```
Comment (jaideepr97) [inline src/instructlab/sdg/utils/chunkers.py:220]: here instead of erroring out we can check if `docling_model_path` is `None` and if it is, then perform the model download here

Comment (jaideepr97) [reply src/instructlab/sdg/utils/chunkers.py:220]: see https://github.com/instructlab/sdg/pull/362/files#r1837291550

File: src/instructlab/sdg/generate_data.py [hunk 4]
```diff
@@ -220,6 +220,11 @@ def _sdg_init(ctx, pipeline):
     data_dirs = [os.path.join(xdg_data_home(), "instructlab", "sdg")]
     data_dirs.extend(os.path.join(dir, "instructlab", "sdg") for dir in xdg_data_dirs())
 
+    # Initialize docling model path
+    docling_model_path = os.path.join(xdg_data_home(), "models", "docling")
+    # Ensure the `docling_model_path` directory exists
+    os.makedirs(docling_model_path, exist_ok=True)
+
```
Comment (jaideepr97) [inline src/instructlab/sdg/generate_data.py:227]: a few suggestions:
- the models themselves will be in `/usr/share/instructlab/sdg/sdg-models/docling-models`
- we should check for the models directory across all dirs populated in `data_dirs`
- we don't need to make sure the directory exists. If it does we proceed, if it does not then we can just set the docling_model_path to be `None` and let it get handled later on 
- what we really need to do is look for a `config.yaml` at `/usr/share/instructlab/models/sdg-models` that will contain the path where the actual models are stored. That's the path we need to pass ahead if it exists in the file

something along the lines of:

```suggestion
    sdg_models_path = docling_models_path = None 
    for d in data_dirs:
        if os.path.exists(os.path.join(d, "models")):
          sdg_models_path = os.path.join(d, "models")
          break
     
     if sdg_models_path is not None:
       try:
          with open(os.path.join(sdg_models_path, "config.yaml"), "r", encoding="utf-8") as file:
              config = yaml.safe_load(file)
              docling_models_path = config['models'][0]['path']
        except (FileNotFoundError, NotADirectoryError, PermissionsError) as e:
              log.warning(f"unable to read docling models path from config.yaml")
```

so we know after this point `docling_models_path` is either set correctly, or it is `None`

File: src/instructlab/sdg/utils/chunkers.py [hunk 5]
```diff
@@ -211,8 +216,17 @@ def chunk_documents(self) -> List:
...
+            logger.info(
+                f"Docling model path {self.docling_model_path} not found, downloading models..."
+            )
+            os.makedirs(self.docling_model_path, exist_ok=True)
+            self.docling_model_path = StandardPdfPipeline.download_models_hf()
+        else:
+            logger.info("Found the docling models")
+
+        pipeline_options = PdfPipelineOptions(artifacts_path=self.docling_model_path)
+
```
Comment (jaideepr97) [inline src/instructlab/sdg/utils/chunkers.py:229]: Here we can just check if supplied coling_model_path is None or not 
there's probably also no point in ensuring what we supplied exists if it is None so we can just use whatever path is returned to us from `StandardPdfPipeline.download_,models_hf()`

```suggestion
        if self.docling_model_path is None:
            logger.info(
                f"Docling models not found on disk, downloading models..."
            )
            self.docling_model_path = StandardPdfPipeline.download_models_hf()
        else:
            logger.info("Found the docling models")

        pipeline_options = PdfPipelineOptions(artifacts_path=self.docling_model_path)

```

File: src/instructlab/sdg/generate_data.py (hunk 1 above)
Comment (jaideepr97) [reply src/instructlab/sdg/generate_data.py]: addressed

File: src/instructlab/sdg/generate_data.py (hunk 2 above)
Comment (jaideepr97) [reply src/instructlab/sdg/generate_data.py]: addressed

File: src/instructlab/sdg/utils/chunkers.py (hunk 3 above)
Comment (jaideepr97) [reply src/instructlab/sdg/utils/chunkers.py]: addressed

**Review by jaideepr97 — APPROVED** [review]
LGTM, thanks

**Review by bbrowning — CHANGES_REQUESTED** [review]
A couple of notes where I don't think the path is actually getting looked up and used like we want. And, as a more general note, I wonder if it would be easier to just lookup the docling models using the xdg_data_dirs logic closer to where we use docling, inside `ContextAwareChunker` as that would avoid us having to wire this model path all the way through the calls here to get it from `generate_data` down to where it's used.

File: src/instructlab/sdg/generate_data.py [hunk 6]
```diff
@@ -220,6 +221,25 @@ def _sdg_init(ctx, pipeline):
     data_dirs = [os.path.join(xdg_data_home(), "instructlab", "sdg")]
     data_dirs.extend(os.path.join(dir, "instructlab", "sdg") for dir in xdg_data_dirs())
 
+    docling_model_path = None
+    sdg_models_path = docling_model_path
+    for d in data_dirs:
+        if os.path.exists(os.path.join(d, "models")):
+            sdg_models_path = os.path.join(d, "models")
+            break
```
[the remark this answers is not in the document]
Comment (jaideepr97) [reply src/instructlab/sdg/generate_data.py:229]: that's probably just an indentation issue. The lines following the break don't need to be in the loop

File: src/instructlab/sdg/generate_data.py [hunk 7]
```diff
@@ -295,6 +315,7 @@ def generate_data(
     batch_size: Optional[int] = None,
     checkpoint_dir: Optional[str] = None,
     max_num_tokens: Optional[int] = DEFAULT_MAX_NUM_TOKENS,
+    docling_model_path: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (jaideepr97) [reply src/instructlab/sdg/generate_data.py:318]: no, users don't need to know about this, it will get propogated as `None` until the last step where there is a check for `none` which triggers a download from HF

File: src/instructlab/sdg/generate_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aakankshaduggal) [reply src/instructlab/sdg/generate_data.py:229]: Yeah, my bad that got indented. Fixing that in my next commit.

File: src/instructlab/sdg/generate_data.py [hunk 8]
```diff
@@ -220,6 +221,25 @@ def _sdg_init(ctx, pipeline):
...
+            sdg_models_path = os.path.join(d, "models")
+            break
+
+        if sdg_models_path is not None:
+            try:
+                with open(
+                    os.path.join(sdg_models_path, "config.yaml"), "r", encoding="utf-8"
+                ) as file:
+                    config = yaml.safe_load(file)
+                    docling_model_path = config["models"][0]["path"]
```
[the remark this answers is not in the document]
Comment (aakankshaduggal) [reply src/instructlab/sdg/generate_data.py:237]: It made sense to have it here, but I hear you. Returning more values than this will be too hacky.

[the remark this answers is not in the document]
Comment (aakankshaduggal) [reply src/instructlab/sdg/generate_data.py:237]: @bbrowning would you suggest changing this flow and eliminate any dependencies on the generate_data.py and rather have all of this as a part of chunkers.py? Since the `chunk_documents` function is being called as a part of the taxonomy functions it made more sense to have this flow instead.

File: src/instructlab/sdg/generate_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaideepr97) [reply src/instructlab/sdg/generate_data.py:229]: good catch @bbrowning ! i have a hard time eye balling indentation in the github diff view

File: src/instructlab/sdg/generate_data.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:318]: Hmm, let me rephrase. Do we expect anyone to ever pass in a path here, bypassing our logic to automatically find the path? Or is it just an oversight that we're adding this as an additional parameter a user could pass in the `generate_data`?

File: src/instructlab/sdg/generate_data.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bbrowning) [reply src/instructlab/sdg/generate_data.py:237]: I think if get the indentation fixed and this value returned from the method so that it gets used by where `_sdg_init` is called, we can clean these other things up downstream. I'm not opposed to working on that now, but also more than happy to revisit this once we get the initial bits working.

[the remark this answers is not in the document]
Comment (aakankshaduggal) [reply src/instructlab/sdg/generate_data.py:237]: I think by doing this we can eliminate the need of passing this as a parameter to generate_data. We just need this value to be passed through the taxonomy functionalities as well

**Comment by bbrowning** [comment]
Changed how my unit tests mock to avoid a Python 3.10 vs 3.11 incompatibility with patching certain modules. This should get the tests green on the Python 3.10 CI again.

**Comment by aakankshaduggal** [comment]
Thanks @bbrowning for adding the quick unit tests 💯

**Review by jaideepr97 — APPROVED** [review]
LGTM 
thanks @aakankshaduggal and @bbrowning ! 
might want to squash your commits before merging

**Review by khaledsulayman — APPROVED** [review]

**Comment by nathan-weinberg** [comment]
@Mergifyio backport release-v0.5
